### PR TITLE
resolve issue#19051

### PR DESCRIPTION
**Pull request by shrutimistry**
#### What does this implement/fix? Explain your changes.

Adds: https://github.com/scikit-learn/scikit-learn/issues/19051

This feature allows the users to specify a font-size for the values of the matrix in the graph that is displayed. The class ConfusionMatrixDisplay is used to plot the confusion matrix visualization for the given confusion matrix along with the display labels. Our change adds the ability to initialize the ConfusionMatrixDisplay with a new `font_size` parameter i.e the class now has a `self.font_size` property. In addition, ConfusionMatrixDisplay also has a setter and getter for the property. Within the plot() function in ConfusionMatrixDisplay, it will create the matrix display with the font size provided by the user otherwise it will use the default font size of 10 (if the font_size provided by the user is None).

Files modified: lets-git-it-scikit-learn/sklearn/metrics/_plot/confusion_matrix.py
Files added to display this Matrix: lets-git-it-scikit-learn/sklearn/bu3.py

### CODE REVIEW

**Comment by shrutimistry** [comment]
@VishwaP98 @linjun9 @RylandWang can you take a look at this when you guys have the time

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 1]
```diff
@@ -77,6 +77,12 @@ def __init__(self, confusion_matrix, *, display_labels=None):
         self.confusion_matrix = confusion_matrix
         self.display_labels = display_labels
 
+    def set_fontsize(self, font_size=None):
+        self.font_size = font_size
+
+    def get_fontsize(self):
+        return self.font_size
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py]: I think it's recommended to use `@property` to implement getter and setter in python.

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 2]
```diff
@@ -77,6 +77,12 @@ def __init__(self, confusion_matrix, *, display_labels=None):
         self.confusion_matrix = confusion_matrix
         self.display_labels = display_labels
 
+    def set_fontsize(self, font_size=None):
+        self.font_size = font_size
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:79]: If we are gonna implement a setter, I think it's better to check the input value here

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 3]
```diff
@@ -142,10 +148,16 @@ def plot(self, *, include_values=True, cmap='viridis',
...
+                if(self.font_size):
+                    self.text_[i, j] = ax.text(
+                        j, i, text_cm,
+                        ha="center", va="center",
+                        color=color, fontsize=self.font_size)
+                else:
+                    self.text_[i, j] = ax.text(
+                        j, i, text_cm,
+                        ha="center", va="center",
+                        color=color)
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py]: what will happen if we set the `fontsize=None` here?

**Review by linjun9 — COMMENTED** [review]
added some comments. It would be great if you can add some screenshots, so we know it's working.

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 1 above)
Comment (linjun9) [reply sklearn/metrics/_plot/confusion_matrix.py]: if you want to take a look, https://www.programiz.com/python-programming/property

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 2 above)
Comment (shrutimistry) [reply sklearn/metrics/_plot/confusion_matrix.py:79]: like for integer type? else - throw error?

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 3 above)
Comment (shrutimistry) [reply sklearn/metrics/_plot/confusion_matrix.py]: if `fontsize=None` then its not gonna pass in font-size to ax.text() it just takes in the default fontsize given from ax.text

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 1 above)
Comment (shrutimistry) [reply sklearn/metrics/_plot/confusion_matrix.py]: okkk will change it, thankss

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 2 above)
Comment (linjun9) [reply sklearn/metrics/_plot/confusion_matrix.py:79]: emm, I think we can check `type` and `font_size > 0`. and yes, throw error if getting invalid input

File: sklearn/metrics/_plot/confusion_matrix.py (hunk 3 above)
Comment (linjun9) [reply sklearn/metrics/_plot/confusion_matrix.py]: so there is no need for this `if else` block, we can do the following code directly?
```
self.text_[i, j] = ax.text(
                        j, i, text_cm,
                        ha="center", va="center",
                        color=color, fontsize=self.font_size)

```

**Comment by shrutimistry** [comment]
@linjun9 resolved the comments and added the picture of the output on our a2 doc: https://docs.google.com/document/d/1n79ByZ6n-7-_O1OspcqcuNeakc91YYtaEibEUIDjCUo/edit

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 4]
```diff
@@ -166,6 +187,9 @@ def plot(self, *, include_values=True, cmap='viridis',
         self.figure_ = fig
         self.ax_ = ax
         return self
+    
+    #creating a property object
+    # font_size = property(get_fontsize, set_fontsize)
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:192]: same

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 5]
```diff
@@ -130,6 +141,8 @@ def plot(self, *, include_values=True, cmap='viridis',
             # print text with appropriate color depending on background
             thresh = (cm.max() + cm.min()) / 2.0
 
+
+
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:145]: unnecessary empty lines

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 6]
```diff
@@ -73,9 +73,20 @@ class ConfusionMatrixDisplay:
     >>> disp.plot() # doctest: +SKIP
     """
     @_deprecate_positional_args
-    def __init__(self, confusion_matrix, *, display_labels=None):
+    def __init__(self, confusion_matrix, *, display_labels=None, font_size=10):
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:76]: you might want to update the doc above.

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 7]
```diff
@@ -142,10 +155,18 @@ def plot(self, *, include_values=True, cmap='viridis',
                 else:
                     text_cm = format(cm[i, j], values_format)
 
+                #set the values of the matrix to the given font size
+                #otherwise, set the values of the matrix to the default font size
+                # if(self.font_size):
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:160]: remove comments, please.

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 8]
```diff
@@ -142,10 +155,18 @@ def plot(self, *, include_values=True, cmap='viridis',
...
                 self.text_[i, j] = ax.text(
                     j, i, text_cm,
                     ha="center", va="center",
-                    color=color)
+                    color=color, fontsize=self._font_size)
+                # else:
+                #     self.text_[i, j] = ax.text(
+                #         j, i, text_cm,
+                #         ha="center", va="center",
+                #         color=color)
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:169]: same.

File: sklearn/metrics/_plot/confusion_matrix.py [hunk 9]
```diff
@@ -553,7 +577,8 @@ def plot_confusion_matrix(estimator, X, y_true, *, labels=None,
             display_labels = labels
 
     disp = ConfusionMatrixDisplay(confusion_matrix=cm,
-                                  display_labels=display_labels)
+                                  display_labels=display_labels,font_size=font_size)
+
```
Comment (linjun9) [inline sklearn/metrics/_plot/confusion_matrix.py:581]: This method has been deprecated. Why do we want to update this?

Comment (shrutimistry) [reply sklearn/metrics/_plot/confusion_matrix.py:581]: It uses it in the issue description to reproduce to bug

**Comment by shrutimistry** [comment]
@linjun9 As discussed, all the fixes for the review comments were made in this new branch: https://github.com/VishwaP98/lets-git-it-scikit-learn/tree/fontsize_soln which needs to be review so that it can merge into this branch, which will act as the main branch for our feature.

TLDR: https://github.com/VishwaP98/lets-git-it-scikit-learn/pull/18  and https://github.com/VishwaP98/lets-git-it-scikit-learn/pull/12 needs to be reviewed first.
